rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1080 @@
+- Specify a different format.
+- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
+- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
+  However, such an approach would not reduce the amount of merge conflicts,
+  and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+- Should the line length limit be specified?
+
+## Future work
+
+- General style guidelines beyond AST reformatting
+- Making widespread use of linters like Nixpkgs-hammering
+- Enforcing the format to other official repositories
+
+-----
+
+## Initial standard Nix format
+
+Terms and definitions:
+- Brackets: `[]`
+- Braces: `{}`
+- Parentheses: `()`
+
+- Newlines are not guaranteed to be preserved, but empty lines are.
+  - This allows the formatter to compact down multi-line expressions if necessary, while still allowing to structure the code appropriately.
+- Expressions of the same kind that can be treated as a sequence of expressions on the same level should be treated as such, even though they are technically parsed as a nested tree.
+  - This applies to else-if chains, functions with multiple arguments, some operators, etc.
+  - Example:
+    ```nix
+    # This is treated as a sequence of if-then-elsa's chains
+    if cond1 then
+      foo
+    else if cond2 then
+      bar
+    else
+      baz
+    ```
+
+### Single-line common ancestor expression rule
+
+For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.
+
+**Example**
+
+```nix
+# Bad, expressions cond and bar are fully on the same line,
+# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
+if cond then foo
+else bar
+
+# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
+# which is also fully on the same line
+if cond then foo else bar
+
+# Bad, due to function application precedence, the smallest common ancestor expression
+# of foo and bar is `foo || bar baz`, which spans two lines
+foo || bar
+  baz
+```
+
+**Rationale**
+
+This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.
+
+### Line length
+
+Lines have a soft length limit that doesn't count indentation.
+There may also be a hard length limit that includes indentation.
+String-like values such as strings, paths, comments, urls, etc. may go over the hard length limit.
+These limits should be configurable, and the soft length limit should default to 100 characters.
+
+### Indentation
+
+- Two spaces are used for each indentation level.
+  - This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.
+- Vertical alignment may not be persisted, neither at the start of the line nor within lines.
```

Comment:
Does ‘may not’ here mean that complying with the standard Nix format *requires* that vertical alignment spaces be replaced with single spaces (or line breaks and indentation), or that the standard Nix format leaves it up to the formatter whether to preserve vertical alignment?

Reply:
Good point, we should make sure to adhere to [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119) to be unambiguous.

Reply:
Vertical alignment produces diff noise. Potentially a lot.

Reply:
Regarding diff noise: we don't want to have our formatter provide "view" and "diff" mode, right? Because we can configure git to use a specific diff format even if the files are formatted differently. The goal is to explicitly **not** have that, yes?

Reply:
A view mode was discussed, but no conclusion was reached IIRC. Personally I wouldn't generally rule it out, but say that it's out of scope for this RFC. So things need to work sufficiently well without for now.

Reply:
Maybe should be the case to add _[disable tags (comments)](https://astyle.sourceforge.net/astyle.html#_Disable_Formatting)_, do we plan to have any?

Subjective comment, I think nothing add more value to ~readability~ navigation of the code than vertical alignment (indentation included),

Not so subjective comment, I found data saying it has no difference.  [A Systematic Literature Review on the Impact of Formatting Elements on Code Legibility](https://arxiv.org/abs/2208.12141) (page 24, table 6)

Reply:
I think such formatting tags should be out of scope for this RFC.

@hugosenari: Nice find!

Reply:
Discussed this today with the shepherd team. We still don't think that there should be vertical alignment, though we generally think it would be okay to change the behavior with special formatting directives, though that is out of scope for this RFC.